// client/src/ui/Chart/index.js
import zingchart from "../../../node_modules/zingchart/es6.js";
import { ZC } from "../../../node_modules/zingchart/es6.js";
import "../../../node_modules/zingchart/modules-es6/zingchart-pareto.min.js";

let ChartView = {};

ChartView.render = function (Sales, Rentals) {
  
// crétion du select
let genres = ["All", "Action", "Animation", "Comedy", "Drama", "Horror", "Romance", "Sci-Fi", "Thriller"];

// id du select
let select = document.createElement("select");
select.id = "GenreSelect";

// style du select
select.style.margin = "10px 10px";
select.style.overflow = "visible";
select.style.padding = "5px";
select.style.border = "1px solid #ccc";
select.style.borderRadius = "5px";
select.style.backgroundColor = "#f9f9f9";
select.style.fontSize = "14px";
select.style.color = "#333";

// création des options
genres.forEach(genre => {
  let option = document.createElement("option");
  option.value = genre;
  option.text = genre;
  select.appendChild(option);
});

// Action
let actionSales = Sales.filter(item => item.genre === "Action");
let allActionSales = actionSales.map(item => item.sales_count);
let actionRentals = Rentals.filter(item => item.genre === "Action");
let allActionRentals = actionRentals.map(item => item.rentals_count);

// Animation
let animationSales = Sales.filter(item => item.genre === "Animation");
let allAnimationSales = animationSales.map(item => item.sales_count);
let animationRentals = Rentals.filter(item => item.genre === "Animation");
let allAnimationRentals = animationRentals.map(item => item.rentals_count);

// Comedy
let comedySales = Sales.filter(item => item.genre === "Comedy");
let allComedySales = comedySales.map(item => item.sales_count);
let comedyRentals = Rentals.filter(item => item.genre === "Comedy");
let allComedyRentals = comedyRentals.map(item => item.rentals_count);

// Drama
let dramaSales = Sales.filter(item => item.genre === "Drama");
let allDramaSales = dramaSales.map(item => item.sales_count);
let dramaRentals = Rentals.filter(item => item.genre === "Drama");
let allDramaRentals = dramaRentals.map(item => item.rentals_count);

// Horror
let horrorSales = Sales.filter(item => item.genre === "Horror");
let allHorrorSales = horrorSales.map(item => item.sales_count);
let horrorRentals = Rentals.filter(item => item.genre === "Horror");
let allHorrorRentals = horrorRentals.map(item => item.rentals_count);

// Romance
let romanceSales = Sales.filter(item => item.genre === "Romance");
let allRomanceSales = romanceSales.map(item => item.sales_count);
let romanceRentals = Rentals.filter(item => item.genre === "Romance");
let allRomanceRentals = romanceRentals.map(item => item.rentals_count);

// Sci-Fi
let scifiSales = Sales.filter(item => item.genre === "Sci-Fi");
let allScifiSales = scifiSales.map(item => item.sales_count);
let scifiRentals = Rentals.filter(item => item.genre === "Sci-Fi");
let allScifiRentals = scifiRentals.map(item => item.rentals_count);

// Thriller
let thrillerSales = Sales.filter(item => item.genre === "Thriller");
let allThrillerSales = thrillerSales.map(item => item.sales_count);
let thrillerRentals = Rentals.filter(item => item.genre === "Thriller");
let allThrillerRentals = thrillerRentals.map(item => item.rentals_count);

let chart = document.getElementById("chart");
chart.appendChild(select);

  let date = ["2024-06", "2024-07", "2024-08", "2024-09", "2024-10", "2024-11"];

  let chartConfig = {
    type: "bar",
    title: "Vente et location par genre",
    backgroundColor: "#f0f0f0", // Change background color here

    series: [

      { values: allActionSales, text: "Action Sales" },
      { values: allActionRentals, text: "Action Rentals" },

      { values: allAnimationSales, text: "Animation Sales" },
      { values: allAnimationRentals, text: "Animation Rentals" },

      { values: allComedySales, text: "Comedy Sales" },
      { values: allComedyRentals, text: "Comedy Rentals" },

      { values: allDramaSales, text: "Drama Sales" },
      { values: allDramaRentals, text: "Drama Rentals" },

      { values: allHorrorSales, text: "Horror Sales" },
      { values: allHorrorRentals, text: "Horror Rentals" },

      { values: allRomanceSales, text: "Romance Sales" },
      { values: allRomanceRentals, text: "Romance Rentals" },

      { values: allScifiSales, text: "Sci-Fi Sales" },
      { values: allScifiRentals, text: "Sci-Fi Rentals" },

      { values: allThrillerSales, text: "Thriller Sales" },
      { values: allThrillerRentals, text: "Thriller Rentals" },
    ],

    scaleX: {
      // convert text on scale indices
      values: date,
      labels: date,
    },

    legend: {},

    plot: {
      tooltip: {
        visible: false,
      },
      alphaArea: 1,
      animation: {
        delay: 500,
        effect: 'ANIMATION_SLIDE_TOP',
        method: 'ANIMATION_LINEAR',
        sequence: 'ANIMATION_BY_PLOT',
        speed: 700,
      },
      aspect: 'spline',
      contourOnTop: false,
      lineWidth: '2px',
      marker: {
        visible: true,
      },
    },

    plotarea: {
      margin: "dynamic"
    },

    tooltip: {
      text: "%t: %v",
      backgroundColor: "#fff",
      borderColor: "#333",
      borderWidth: 1,
      borderRadius: 5,
      fontColor: "#333"
    },

    crosshairX: {
      plotLabel: {
        backgroundColor: '#434343',
        fontColor: '#FFF',
      },
      scaleLabel: {
        backgroundColor: '#fff',
        borderColor: '#333',
        borderRadius: '3px',
        borderWidth: '1px',
        fontColor: 'black',
      },
    },

    scaleY: {
      label: {
        text: "Count"
      }
    }
  }

  // event listener pour le select par genre et afficher les données correspondantes
  select.addEventListener("change", function() {
    if (select.value === "All") {
      chartConfig.series = [
        { values: allActionSales, text: "Action Sales" },
        { values: allActionRentals, text: "Action Rentals" },

        { values: allAnimationSales, text: "Animation Sales" },
        { values: allAnimationRentals, text: "Animation Rentals" },

        { values: allComedySales, text: "Comedy Sales" },
        { values: allComedyRentals, text: "Comedy Rentals" },

        { values: allDramaSales, text: "Drama Sales" },
        { values: allDramaRentals, text: "Drama Rentals" },

        { values: allHorrorSales, text: "Horror Sales" },
        { values: allHorrorRentals, text: "Horror Rentals" },

        { values: allRomanceSales, text: "Romance Sales" },
        { values: allRomanceRentals, text: "Romance Rentals" },

        { values: allScifiSales, text: "Sci-Fi Sales" },
        { values: allScifiRentals, text: "Sci-Fi Rentals" },

        { values: allThrillerSales, text: "Thriller Sales" },
        { values: allThrillerRentals, text: "Thriller Rentals" },
      ];
    }

    // si le genre est "Action"
    if (select.value === "Action") {
      chartConfig.series = [
        { values: allActionSales, text: "Action Sales" },
        { values: allActionRentals, text: "Action Rentals" },
      ];
    }

    // si le genre est "Animation"
    if (select.value === "Animation") {
      chartConfig.series = [
        { values: allAnimationSales, text: "Animation Sales" },
        { values: allAnimationRentals, text: "Animation Rentals" },
      ];
    }

    // si le genre est "Comedy"
    if (select.value === "Comedy") {
      chartConfig.series = [
        { values: allComedySales, text: "Comedy Sales" },
        { values: allComedyRentals, text: "Comedy Rentals" },
      ];
    }

    // si le genre est "Drama"
    if (select.value === "Drama") {
      chartConfig.series = [
        { values: allDramaSales, text: "Drama Sales" },
        { values: allDramaRentals, text: "Drama Rentals" },
      ];
    }

    // si le genre est "Horror"
    if (select.value === "Horror") {
      chartConfig.series = [
        { values: allHorrorSales, text: "Horror Sales" },
        { values: allHorrorRentals, text: "Horror Rentals" },
      ];
    }

    // si le genre est "Romance"
    if (select.value === "Romance") {
      chartConfig.series = [
        { values: allRomanceSales, text: "Romance Sales" },
        { values: allRomanceRentals, text: "Romance Rentals" },
      ];
    }

    // si le genre est "Sci-Fi"
    if (select.value === "Sci-Fi") {
      chartConfig.series = [
        { values: allScifiSales, text: "Sci-Fi Sales" },
        { values: allScifiRentals, text: "Sci-Fi Rentals" },
      ];
    }

    // si le genre est "Thriller"
    if (select.value === "Thriller") {
      chartConfig.series = [
        { values: allThrillerSales, text: "Thriller Sales" },
        { values: allThrillerRentals, text: "Thriller Rentals" },
      ];
    }

    zingchart.render({
      id: "chart",
      data: chartConfig,
      height: "100%",
      width: "100%",
    });
  });

  zingchart.render({
    id: "chart",
    data: chartConfig,
    height: "100%",
    width: "100%",
  });
};

export { ChartView };
