// client/src/ui/Chart/index.js
import zingchart from "../../../node_modules/zingchart/es6.js";
import { ZC } from "../../../node_modules/zingchart/es6.js";
import "../../../node_modules/zingchart/modules-es6/zingchart-pareto.min.js";

let ChartView = {};

ChartView.render = function (SalesCountry, RentalsCountry) {

// console.log(SalesCountry);
// console.log(RentalsCountry);

// SalesCountry
/*
[
    {
        "country": "Denmark",
        "total_sales": 27.42
    },
    {
        "country": "France",
        "total_sales": 209.67
    },
    {
        "country": "Greece",
        "total_sales": 305.75
    },
    {
        "country": "Portugal",
        "total_sales": 291.1
    },
    {
        "country": "Spain",
        "total_sales": 30.93
    },
    {
        "country": "Sweden",
        "total_sales": 211.37
    },
    {
        "country": "United Kingdom",
        "total_sales": 37.28
    }
]
*/
  
// RentalsCountry
/*
[
    {
        "country": "Austria",
        "total_rentals": 53.2
    },
    {
        "country": "Denmark",
        "total_rentals": 144.52
    },
    {
        "country": "France",
        "total_rentals": 692
    },
    {
        "country": "Germany",
        "total_rentals": 77.58
    },
    {
        "country": "Greece",
        "total_rentals": 691.42
    },
    {
        "country": "Portugal",
        "total_rentals": 957.19
    },
    {
        "country": "Spain",
        "total_rentals": 53
    },
    {
        "country": "Sweden",
        "total_rentals": 800.16
    },
    {
        "country": "United Kingdom",
        "total_rentals": 70.17
    }
]
*/

// crétion du select
let type = ["Sales", "Rentals", "All"];

// id du select
let select = document.createElement("select");
select.id = "GenreSelect";

// style du select
select.style.margin = "10px 10px";
select.style.overflow = "visible";
select.style.padding = "5px";
select.style.border = "1px solid #ccc";
select.style.borderRadius = "5px";
select.style.backgroundColor = "#f9f9f9";
select.style.fontSize = "14px";
select.style.color = "#333";

// création des options
type.forEach(type => {
  let option = document.createElement("option");
  option.value = type;
  option.text = type;
  select.appendChild(option);
});

let chart_select = document.getElementById("div_select");
chart_select.appendChild(select);

// Sales
let allSales = SalesCountry.map(item => item.total_sales);
console.log("Sales :", allSales);
let allCountriesSales = SalesCountry.map(item => item.country);
console.log("Pays de Vente :", allCountriesSales);

// Rentals
let allRentals = RentalsCountry.map(item => item.total_rentals);
console.log("Rentals", allRentals);
let allCountriesRentals = RentalsCountry.map(item => item.country);
console.log("Pays de locations", allCountriesRentals);

// Configuration du graphique
let chartConfig = {
  type: "hbullet",
  title: {
    text: "Vente et location par Pays",
  },
  backgroundColor: "#f0f0f0",

  series: [
    { values: allSales, text: "Sales" },
    { values: allRentals, text: "Rentals" },
  ],

  scaleX: {
    labels: allCountriesRentals,
    item: {
      angle: 45,
      wrapText: true, // Allow text wrapping
      maxChars: 7, // Maximum characters per line
    },
    fontColor: "#000000", // couleur des labels (noir) 
    fontSize: "12px",
  },

  scaleY: {
    label: {
      text: "Total",
    },
  },

  legend: {
    draggable: true,
  },

  plot: {
    animation: {
      effect: 3,
      sequence: 1,
      speed: 700,
    },
    aspect: "spline",
    marker: {
      visible: false,
    },
  },

  tooltip: {
    text: "%t: %v",
    backgroundColor: "#fff",
    borderColor: "#333",
    borderWidth: 1,
    borderRadius: 5,
    fontColor: "#333",
  },

  crosshairX: {
    plotLabel: {
      backgroundColor: '#434343',
      fontColor: '#FFF',
    },

    scaleLabel: {
      backgroundColor: '#fff',
      borderColor: '#333',
      borderRadius: '3px',
      borderWidth: '1px',
      fontColor: 'black',
    },
  },
};

select.addEventListener("change", function() {
  if (select.value === "Sales") {
    chartConfig.series = [
      { values: allSales, text: "Sales" },
    ];
    chartConfig.scaleX.labels = allCountriesSales;
  }

  if (select.value === "Rentals") {
    chartConfig.series = [
      { values: allRentals, text: "Rentals" },
    ];
    chartConfig.scaleX.labels = allCountriesRentals;
  }

  if (select.value === "All") {
    chartConfig.series = [
      { values: allSales, text: "Sales" },
      { values: allRentals, text: "Rentals" },
    ];
    chartConfig.scaleX.labels = [...new Set([...allCountriesSales, ...allCountriesRentals])];
  }

  zingchart.render({
    id: "chart",
    data: chartConfig,
    height: "100%",
    width: "100%",
  });
});

  zingchart.render({
    id: "chart",
    data: chartConfig,
    height: "100%",
    width: "100%",
  });
};

export { ChartView };
